railties/lib/rails/tasks/routes.rake

```diff
@@ -1,7 +1,36 @@
-desc 'Print out all defined routes in match order, with names. Target specific controller with CONTROLLER=x.'
+require 'active_support/deprecation'
+require 'active_support/core_ext/string/strip' # for strip_heredoc
+require 'optparse'
+
+desc 'Print out all defined routes in match order, with names. Target specific controller with --controller option - or its -c shorthand.'
 task routes: :environment do
   all_routes = Rails.application.routes.routes
   require 'action_dispatch/routing/inspector'
   inspector = ActionDispatch::Routing::RoutesInspector.new(all_routes)
-  puts inspector.format(ActionDispatch::Routing::ConsoleFormatter.new, ENV['CONTROLLER'])
+  if ARGV.any?{|argv| argv.start_with? 'CONTROLLER'}
+    puts <<-eow.strip_heredoc
+      Passing `CONTROLLER` to `bin/rake routes` is deprecated and will be removed in Rails 5.1.
+      Please use `bin/rake routes -c controller_name` instead.
+    eow
+  end
+
+  options = { controller: ENV['CONTROLLER'] }
+
+  OptionParser.new do |opts|
+    opts.banner = "Usage: rake routes [options]"
+    opts.on("-c", "--controller [PATTERN]") do |pattern|
+      options[:filter] = pattern
+      options[:defaults_filter] = { controller: /#{pattern.downcase.sub(/_?controller\z/, '').sub('::', '/')}/ }
+    end
+
+    opts.on("-g", "--grep [PATTERN]") do |pattern|
+      options[:filter] = pattern
+      options[:defaults_filter] = ActionDispatch::Routing::RoutesInspector.defaults_filter_for(pattern)
```

Comment:
Feels silly to fill out an options hash, when we don't have to use it. Let's do:

```ruby
routes_filter = { controller: ENV['CONTROLLER'] }

# ...
opts.on("-c", "--controller [CONTROLLER]") do |controller|
  routes_filter = { controller: controller }
end

opts.on("-g", "--grep [PATTERN]") do |pattern|
  routes_filter = pattern
end

# In the inspector...
def normalize_filter(filter)
  if filter.respond_to?(:[]) && filter[:controller]
    filter[:controller] = /#{filter[:controller].downcase.sub(/_?controller\z/, '').sub('::', '/')}/
  else
    { controller: /#{filter}/, action: /#{filter}/ }
  end
end
```

Notice the rename in the `on` calls. The controller argument shouldn't be called pattern, think that should just be called controller.